cmd/e2e-runner/main.go

```diff
@@ -175,26 +177,63 @@ func realMain(ctx context.Context) error {
 }
 
 // handleDefault handles the default end-to-end scenario.
-func handleDefault(cfg *config.E2ERunnerConfig, h *render.Renderer) http.Handler {
+func handleDefault(cfg *config.E2ERunnerConfig, db *database.Database, h *render.Renderer) http.Handler {
 	c := *cfg
 	c.DoRevise = false
-	return handleEndToEnd(&c, h, mDefaultSuccess)
+	c.DoUserReport = false
+	return handleEndToEnd(&c, db, h, mDefaultSuccess)
 }
 
 // handleRevise runs the end-to-end runner with revision tokens.
-func handleRevise(cfg *config.E2ERunnerConfig, h *render.Renderer) http.Handler {
+func handleRevise(cfg *config.E2ERunnerConfig, db *database.Database, h *render.Renderer) http.Handler {
 	c := *cfg
 	c.DoRevise = true
-	return handleEndToEnd(&c, h, mRevisionSuccess)
+	c.DoUserReport = false
+	return handleEndToEnd(&c, db, h, mRevisionSuccess)
+}
+
+// handleUserReport runs the end-to-end runner initiated by a user-report API request.
+func handleUserReport(cfg *config.E2ERunnerConfig, db *database.Database, h *render.Renderer) http.Handler {
+	if !cfg.Features.EnableUserReport {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			ctx := r.Context()
+			logger := logging.FromContext(ctx)
+			logger.Warnw("user-report feature is not enabled on this server")
+			stats.Record(ctx, mUserReportSuccess.M(1))
+			h.RenderJSON(w, http.StatusOK, nil)
+		})
+	}
+
+	c := *cfg
+	c.DoRevise = false
+	c.DoUserReport = true
+	return handleEndToEnd(&c, db, h, mUserReportSuccess)
 }
 
 // handleEndToEnd handles the common end-to-end scenario. m is incremented iff
 // the run succeeds.
-func handleEndToEnd(cfg *config.E2ERunnerConfig, h *render.Renderer, m *stats.Int64Measure) http.Handler {
+func handleEndToEnd(cfg *config.E2ERunnerConfig, db *database.Database, h *render.Renderer, m *stats.Int64Measure) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		ctx := r.Context()
 		logger := logging.FromContext(ctx)
 
+		if cfg.DoUserReport {
+			userReport, err := db.FindUserReport(db.RawDB(), project.TestPhoneNumber)
```

Comment:
This API is a little weird to me. If `db` is the handler, why do we need to pass in a `db` as a param?

Reply:
I refactored this into user_report.go